docs/Quiz Feature Planning.md

```diff
@@ -0,0 +1,311 @@
+
+---
+
+## **Analysis of Current System Structure**
+
+### **Main Backend**
+
+- **Responsibilities:**
+    
+    - Handles user requests from the extension or web interface.
+        
+    - Manages session data and interacts with various services, including AI Service.
+        
+- **Key Components:**
+    
+    - `cmd/`: Entry points for the application, defining how the backend starts and operates.
+        
+    - `pkg/`: Contains packages for routing, data handling, and integration logic.
+
+### **AI Service**
+
+- **Responsibilities:**
+    
+    - Processes AI-specific tasks such as question answering and contextual understanding.
+        
+    - Leverages OpenAI's GPT models to generate responses.
+        
+- **Key Components:**
+    
+    - `cmd/`: Main entry points for AI-specific services.
+        
+    - `pkg/`: Contains core logic for AI interaction, Redis caching, and OpenAI API integration.
+        
+
+### **YouTube Info Service**
+- **Responsibilities:**
+    
+    - Fetches video metadata and transcripts from the YouTube API.
+
+- **Key Components:**
+    
+    - `app/main.py`: Hosts the FastAPI application and includes the routes.
+        
+    - `app/service.py`: Retrieves the transcript along with other info of the video.
+
+    - `app/routes.py`: Exposes an endpoint /video-info/{video_id} to retrieve video metadata and transcripts.
+    
+    ---
+
+## **Implementation Options**
+
+
+### **Option 1: Quiz questions generated at the beginning of each video**
+
+- **Description:**
+
+    -In this approach, the entire quiz is generated at the start of the video based on the full transcript. The quiz will consist of a pre-determined number of questions that cover the key points of the entire video, along with explanation for each option.
+
+- **Advantages:**
+
+    -Existing prototypes, such as quizlet.
+
+    -Easier to implement.
+
+- **Disadvantages:**
+
+    -Processing the entire transcript at once can be computationally expensive.
+
+    -Doesn't adapt to the user’s engagement or focus during the video.
+
+
+### **Option 2: Dynamic Quiz Generation Based on Timestamp**
+
+- **Description:**
+
+    -In this approach, quiz questions are generated dynamically as the user progresses through the video. Questions are created based on the relevant part of the transcript corresponding to the current timestamp.
+
+- **Advantages:**
+
+    -Questions are tailored to the user’s current position in the video, enhancing engagement.
+
+    -Encourages users to pay attention to specific sections of the video.
+
+- **Disadvantages:**
+
+    -Requires real-time processing and integration with timestamp-based logic.
+
+    -logic could be complicated, regarding which amount of transcripts before and after the timestamp being used.
+
+
+
+## **Architecture and Data Flow**
+
+### **API Routes**
```

Comment:
I think we can get rid of most of these routes and implementation since we this will be decided when implementing. Doesnt seem right to pre define these things in an ADR. Maybe you could make it less detailed and just write down in bullet points how it could end up looking instead of defining the whole routes and payloads. Just give smaller bullet points on what data we might pass in places and what data we will hope to get etc.